ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -585,7 +586,49 @@ releaseResources rr keys releaser = do
 --
 -- See documentation of 'ResourceRegistry' for a detailed discussion.
 withRegistry :: (IOLike m, HasCallStack) => (ResourceRegistry m -> m a) -> m a
-withRegistry = bracket unsafeNewRegistry closeRegistry
+withRegistry body =
+    bracketWithRegistry
+      (return (const ()))
+      (\() -> return ())
+      $ \registry () -> body registry
+
+-- | Create a new registry owned by some bracketed resources
+--
+-- Use this combinator to emphasize that the lifetime of the registry exactly
+-- spans the lifetime of the bracketed resources, no more and no less.
+--
+-- Despite the coincident lifetimes, we must take care with the simultaneity at
+-- the endpoints. This combinator ensures the following invariants.
+--
+-- o The registry itself is older than the bracketed resources.
+--
+-- o The bracketed resources are older than everything in the registry.
+--
+-- In particular, this combinator closes the registry before releasing the
+-- bracketed resources, since any objects still in the registry will be younger
+-- than the bracketed resources and so may rely on them in their releasers.
+--
+-- See documentation of 'ResourceRegistry' for a more general discussion.
+bracketWithRegistry ::
+     (IOLike m, HasCallStack)
+  => m (ResourceRegistry m -> a)
+     -- ^ Allocate the bracketed resources. They depend on the registry but are
+     -- not themselves registered.
+  -> (a -> m ())
+     -- ^ Release the bracketed resources.
+  -> (ResourceRegistry m -> a -> m r)
```

Comment:
Edsko clarified: we're removing this argument to the body because this combinator captures the intent of providing a bracketed thing that happens to need a registry. That registry is private. NB if the body also needs a registry, it can allocate its own. NB number 2: a prompt invocation of close for `a` could also close the registry.

Ideally we'd write `withFollower`, and `withFollower` would be defined via `withRegistry`, which would guarantee that it's private. This combinator is intended as a generalization of `withFollower` that can be used with any allocation function that expected a privately owned registry. (But we don't think an `ST` style existential type is worthwhile.)

Something like "This combinator provides a private repository. If the allocator does not leak the registry, then it will remain private."